RoaringBitmap/src/main/java/org/roaringbitmap/ArrayContainer.java

```diff
@@ -1485,9 +1485,14 @@ void wrap(ArrayContainer p) {
   }
 
   @Override
-  public CharIterator clone() {
+  public void advanceIfNeeded(char maxval) {
```

Comment:
To the `PeekableCharIterator` definition, the char parameter is a `minval`. Your implementation have departed from the interface at the reverse iterator situation.  

I think we just need to keep the  `advanceIfNeeded` implementation logic in the `ArrayContainerCharIterator.advanceIfNeeded` at the reverse iterator situation.
This advice also applies to other containers.

Reply:
The `advanceIfNeeded` method needs to be in shared interface across all the container for the long bitset reverse calculation.

Would you prefer updating the documentation on the `PeekableCharIterator`
New interfaces i.e. `PeekableForwardCharIterator` and `PeekableReverseCharIterator` with more focused documentation that both extend `PeekableCharIterator`

Or any other suggestion.

For our use case having the same interface for `PeekableLongIterator` regardless of direction would be beneficial.

Reply:
I think we don't need to reimplement the `advanceIfNeeded` method in the reverse direction. The forward and reverse iterator could share a common abstract class which implement the `PeekableCharIterator`.

Reply:
I don't think I quite understand sorry. 
When its a reverse iterator we want to use advanceIfNeeded to skip calls to next. The advanceIfNeeded logic in the reverse iterator cases is flipped. I'm struggling to see how to merge these into a common class

Did update the javadoc in the `PeekableCharIterator` to try reflect this.

Reply:
I would think that advanceIfNeeded would move in reverse and so maxval looks good to me... 

???

Reply:
That would be a bad interface definition to `advanceIfNeeded` to have different input parameter meaning at different implementations.

Reply:
To my understanding, the `advanceIfNeeded` is to find a demarcation point position in a forward number sequence by the input parameter. At that position, the forward numbers would be smaller than the input parameter, the backward numbers would be larger than the input parameter. `PeekableCharIterator` is just to find this position by the `advanceIfNeeded` method.

The reverse or forward iterator's  `next` method implementation define its direction not by the `advanceIfNeeded`'s `maxvalue` or `minvalue`.

So I assume the reverse or forward iterator could extend a common class which implement the `PeekableCharIterator`.

I am not the author of the `PeekableCharIterator`. If I am wrong, please correct me.

Reply:
My interpretation was that the `advanceIfNeeded` method is present to allow efficient skipping through the iterator to prevent the need to call `next` constantly. So in reverse it would be maxval. This is how we plan to use. If it shares the same interface then our implementing code will be simpler for forward and reverse. 

Happy to implement in which ever way you want though

Reply:
I mean , to one container, we just need to implement the `PeekableCharIterator.advanceIfNeeded ` according to only one rule (the current `minval` rule) to find the position to advance to. Maybe the input parameter could be called `thresholdVal`.

Reply:
have changed the name to `thresholdVal` and updated the javadoc in `PeekableCharIterator` and `PeekableLongIterator`